### PR TITLE
accept direct dialer

### PR DESCRIPTION
**Pull request by joesis**
For https://github.com/getlantern/lantern-internal/issues/924. The direct dialer is passed in, instead of being hardcoded as `netx.Dial`. The package becomes context aware too, so the direct dialer can selectively override timeout in different situations.

### CODE REVIEW

File: detour.go [hunk 1]
```diff
@@ -124,19 +120,22 @@ func SetCountry(country string) {
...
+		dc := &Conn{dialDetour: detourDialer, network: network, addr: addr}
 		if !whitelisted(addr) {
 			log.Tracef("Attempting direct connection for %v", addr)
 			detector := blockDetector.Load().(*Detector)
 			dc.setState(stateInitial)
-			// always try direct connection first
-			dc.conn, err = netx.DialTimeout(network, addr, TimeoutToDetour)
+			// Always try direct connection first. The caller may choose a
+			// deadline shorter than the context passed in.
+			dc.conn, err = directDialer(ctx, network, addr)
```
Comment (joesis) [inline detour.go]: This is the main change.

**Comment by joesis** [comment]
Hey sorry @uaalto could you have a look at this too? It's a dependence of https://github.com/getlantern/flashlight/pull/277

**Comment by myleshorton** [comment]
Looks like this has actually been in use for awhile -- merging

**Comment by joesis** [comment]
Oh I totally forgot it. Thanks @myleshorton !

**Pull request merged**